This project is from the DataQuest course "Pandas and NumPy Fundamentals". It is semi-guided with general prompts. The dataset for this project comes from used car listings in the classifieds section of the German eBay website. First this data needs to be cleaned and then we will analyze it to look for interesting patterns.

Here are the column names and descriptions:
* `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
* `name` - Name of the car.
* `seller` - Whether this seller is private or a dealer.
* `offerType` - The type of listing
* `price` - The price on the ad to sell the car
* `abtest` - Whether the listing is included in an A/B test.
* `vehicleType` - The vehicle type
* `yearOfRegistration` - The year in which the car was first registered.
* `gearbox` - The transmission type.
* `powerPS` - The power of the car in PS.
* `model` - The car model name
* `kilometer` - How many kilometers the car has driven.
* `monthOfRegistration` - The month in which the car was first registered.
* `fuelType` - What type of fuel the car uses.
* `brand` - The brand of the car.
* `notRepairedDamage` - If the car has damage which is not yet repaired.
* `dateCreated` - The date on which the eBay listing was created.
* `nrOfPictures` - The number of pictures in the ad.
* `postalCodes` - The postal code for the location of the vehicle.
* `lastSeenOnline` - When the crawler saw this ad last online.

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

A few observations:
* there do no appear to be a large number of missing values, but about 20% of values are missing in `notRepairedDamage`
* car names have underscores instead of spaces
* column titles are calmelcase

In order to make things easier, we are first going to change the names of a few columns and then make all column names lowercase.

In [4]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [5]:
autos.rename(columns={'yearOfRegistration':'registration_year',
                      'monthOfRegistration':'registration_month',
                      'notRepairedDamage':'unrepaired_damage',
                      'dateCreated':'ad_created'},   
             inplace=True)
autos.columns = [x.lower() for x in autos.columns]
autos.head()

datecrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offertype   price   abtest vehicletype  registration_year  \
0  privat   Angebot  $5,000  control         bus               2004   
1  privat   Angebot  $8,500  control   limousine               1997   
2  privat   Angebot  $8,990     test   limousine               2009   
3  privat   Angebot  $4,350  control  kleinwagen               2007   
4  privat   Angebot  $1,350     test       kombi               2003   

     gearbox  powerps   model   odometer  registration_month fueltype  \
0    manuell      158  andere  150,000km                   3      lpg   
1  automatik      286     7er  150,000km                   6   benzin   
2    manuell      102    golf   70,000km                   7   benzin   
3  automatik       71  fortwo   70,000km                   6   benzin   
4    manuell        0   focus  150,000km                   7   benzin   

        brand unrepaired_damage           ad_created  nrofpictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalcode             lastseen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Now we're going to start exploring the data. Let's first look for columns where the values are all or almost all the same-- they won't give us any useful information for our analyses. 

In [6]:
autos.describe(include='all')

datecrawled         name  seller offertype  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-16 21:50:53  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicletype  registration_year  gearbox       powerps  model  \
count        44905       50000.000000    47320  50000.000000  47242   
unique           8                NaN        2           NaN    245   
top      limousine                NaN  manuell           NaN   golf   
freq         12859                NaN    36993           NaN   4024   
mean           NaN        2005.073280      NaN    116.355920    NaN   
std            NaN         105.712813      NaN    209.216627    NaN   
min            NaN        1000.000000      NaN      0.000000    NaN   
25%            NaN        1999.000000      NaN     70.000000    NaN   
50%            NaN        2003.000000      NaN    105.000000    NaN   
75%            NaN        2008.000000      NaN    150.000000    NaN   
max            NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fueltype       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  nrofpictures    postalcode             lastseen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

It looks like:
* almost all sellers are private
* almost all offer types are Angebot
* there are no pictures for any of these ads
* price is a string
* odometer value is a string and includes "km"

We can verify a few of these below:

In [7]:
autos["seller"].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [8]:
autos["offertype"].value_counts()

Angebot    49999
Gesuch         1
Name: offertype, dtype: int64

Let's drop the `seller` and `offertype` columns, convert `price` and `odometer` to numeric types, and then finally rename `odometer` to `odometer_km`.

In [9]:
autos.drop('offertype', axis=1)
autos.drop('seller', axis=1)

datecrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [10]:
autos['price'] = autos['price'].str.replace('$', '').str.replace(',','').astype(float)
autos['odometer'] = autos['odometer'].str.replace('km', '').str.replace(',', '').astype(float)
autos.rename(columns={'odometer':'odometer_km'},   
             inplace=True)

In [11]:
autos.dtypes

datecrawled            object
name                   object
seller                 object
offertype              object
price                 float64
abtest                 object
vehicletype            object
registration_year       int64
gearbox                object
powerps                 int64
model                  object
odometer_km           float64
registration_month      int64
fueltype               object
brand                  object
unrepaired_damage      object
ad_created             object
nrofpictures            int64
postalcode              int64
lastseen               object
dtype: object

Now let's look a little closer at the odometer and price values.

In [12]:
odometer_val = autos['odometer_km']
print(odometer_val.unique().shape)
print(odometer_val.describe())
print(odometer_val.value_counts().sort_index(ascending=True).head())

(13,)
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
5000.0     967
10000.0    264
20000.0    784
30000.0    789
40000.0    819
Name: odometer_km, dtype: int64


In [13]:
price_val = autos['price']
print(price_val.unique().shape)
print(price_val.describe())
print(price_val.value_counts().sort_index(ascending=False).head(15))

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
Name: price, dtype: int64


There do not seem to be any obvious outliers for the odometer column-- all are between 5,000 and 150,000 which are possible values for a car. However, there are a few outliers for the price column. In particular, there are values where the price seems unreasonably high. Given that the 75th percentile is at 7.5K, all of the values in the millions are unlikely. We will remove everything above $350K because that seems to be the last value before a dramatic increase.

In [14]:
autos = autos[autos["price"] < 350000]

In [15]:
price_val = autos['price']
print(price_val.unique().shape)
print(price_val.describe())
print(price_val.value_counts().sort_index(ascending=False).head(5))

(2346,)
count     49985.000000
mean       5714.637531
std        8850.742575
min           0.000000
25%        1100.000000
50%        2950.000000
75%        7200.000000
max      345000.000000
Name: price, dtype: float64
345000.0    1
299000.0    1
295000.0    1
265000.0    1
259000.0    1
Name: price, dtype: int64


Now we're going to look at the dates. Let's start with `datecrawled`, `ad_created`, and `last_seen` because these are all currently just strings in our dataset.

In [16]:
print(autos['datecrawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index())

2016-03-05    0.025388
2016-03-06    0.013944
2016-03-07    0.035971
2016-03-08    0.033270
2016-03-09    0.033210
2016-03-10    0.032130
2016-03-11    0.032490
2016-03-12    0.036771
2016-03-13    0.015565
2016-03-14    0.036631
2016-03-15    0.033990
2016-03-16    0.029509
2016-03-17    0.031509
2016-03-18    0.013064
2016-03-19    0.034910
2016-03-20    0.037831
2016-03-21    0.037491
2016-03-22    0.032910
2016-03-23    0.032390
2016-03-24    0.029109
2016-03-25    0.031750
2016-03-26    0.032490
2016-03-27    0.031029
2016-03-28    0.034850
2016-03-29    0.034150
2016-03-30    0.033630
2016-03-31    0.031910
2016-04-01    0.033810
2016-04-02    0.035411
2016-04-03    0.038692
2016-04-04    0.036491
2016-04-05    0.013104
2016-04-06    0.003181
2016-04-07    0.001420
Name: datecrawled, dtype: float64


It looks like this dataset was collected between March 5, 2016 and April 7, 2016. There don't appear to be any spikes, there are a similar number pulled each day.

In [17]:
print(autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index())

2015-06-11    0.000020
2015-08-10    0.000020
2015-09-09    0.000020
2015-11-10    0.000020
2015-12-05    0.000020
2015-12-30    0.000020
2016-01-03    0.000020
2016-01-07    0.000020
2016-01-10    0.000040
2016-01-13    0.000020
2016-01-14    0.000020
2016-01-16    0.000020
2016-01-22    0.000020
2016-01-27    0.000060
2016-01-29    0.000020
2016-02-01    0.000020
2016-02-02    0.000040
2016-02-05    0.000040
2016-02-07    0.000020
2016-02-08    0.000020
2016-02-09    0.000040
2016-02-11    0.000020
2016-02-12    0.000060
2016-02-14    0.000040
2016-02-16    0.000020
2016-02-17    0.000020
2016-02-18    0.000040
2016-02-19    0.000060
2016-02-20    0.000040
2016-02-21    0.000060
                ...   
2016-03-09    0.033230
2016-03-10    0.031870
2016-03-11    0.032790
2016-03-12    0.036611
2016-03-13    0.016925
2016-03-14    0.035231
2016-03-15    0.033750
2016-03-16    0.030009
2016-03-17    0.031189
2016-03-18    0.013724
2016-03-19    0.033850
2016-03-20    0.037871
2016-03-21 

The ads created for these listings were created between June 2015 and April 2016, telling us that some of these cars have been listed for a long time and some are brand new.

In [18]:
print(autos['lastseen'].str[:10].value_counts(normalize=True, dropna=False).sort_index())

2016-03-05    0.001080
2016-03-06    0.004421
2016-03-07    0.005362
2016-03-08    0.007582
2016-03-09    0.009843
2016-03-10    0.010763
2016-03-11    0.012524
2016-03-12    0.023807
2016-03-13    0.008983
2016-03-14    0.012804
2016-03-15    0.015885
2016-03-16    0.016445
2016-03-17    0.027928
2016-03-18    0.007422
2016-03-19    0.015745
2016-03-20    0.020706
2016-03-21    0.020726
2016-03-22    0.021586
2016-03-23    0.018586
2016-03-24    0.019566
2016-03-25    0.019206
2016-03-26    0.016965
2016-03-27    0.016005
2016-03-28    0.020846
2016-03-29    0.022327
2016-03-30    0.024847
2016-03-31    0.023827
2016-04-01    0.023107
2016-04-02    0.024887
2016-04-03    0.025368
2016-04-04    0.024627
2016-04-05    0.124277
2016-04-06    0.220986
2016-04-07    0.130959
Name: lastseen, dtype: float64


These values seem to mirror the `datecrawled` column.

In [19]:
print(autos['registration_year'].describe())

count    49985.000000
mean      2005.075503
std        105.728207
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64


Most of these cars seem to be ~ 10 years old (2003 to 2008). There are a few obvious outliers though-- values like 1000 and 9999 do not make sense in this context.

In [20]:
print(autos['registration_year'].between(1900,2016).sum())

48015


There are about 2,000 cars that fall outside the range of 1900-2016.

In [21]:
print(autos['registration_year'].sort_values().head(25))
print(autos['registration_year'].sort_values().tail(5))

22316    1000
49283    1001
24511    1111
35238    1500
32585    1800
10556    1800
3679     1910
30781    1910
22659    1910
45157    1910
42181    1910
28693    1910
15898    1910
46213    1910
33295    1910
21416    1927
22101    1929
11246    1931
2573     1934
2221     1934
21421    1937
39725    1937
26607    1937
23804    1937
26103    1938
Name: registration_year, dtype: int64
6308     9996
38076    9999
14341    9999
8012     9999
33950    9999
Name: registration_year, dtype: int64


The range of 1900 to 2016 seems like a reasonable range of cut off values. A car cannot be listed in April 2016 and be a 2017 vehicle (later in the year possibly). The Ford Model T debuted in 1908 and the next value below that is 1800 which would be impossible.

In [22]:
autos = autos[autos["registration_year"].between(1900,2016)]

In [23]:
print(autos["registration_year"].value_counts(normalize=True))

2000    0.069853
2005    0.062793
1999    0.062439
2004    0.057003
2003    0.056795
2006    0.056378
2001    0.056274
2002    0.052754
1998    0.051088
2007    0.047985
2008    0.046465
2009    0.043674
1997    0.042237
2011    0.034031
2010    0.033260
1996    0.030074
2012    0.027554
2016    0.027387
1995    0.027325
2013    0.016786
2014    0.013850
1994    0.013746
1993    0.009268
2015    0.008310
1990    0.008227
1992    0.008122
1991    0.007414
1989    0.003770
1988    0.002957
1985    0.002166
          ...   
1966    0.000458
1977    0.000458
1975    0.000396
1969    0.000396
1965    0.000354
1964    0.000250
1910    0.000187
1963    0.000187
1959    0.000146
1961    0.000125
1956    0.000104
1958    0.000083
1937    0.000083
1962    0.000083
1950    0.000062
1954    0.000042
1941    0.000042
1934    0.000042
1957    0.000042
1951    0.000042
1955    0.000042
1931    0.000021
1953    0.000021
1943    0.000021
1938    0.000021
1939    0.000021
1927    0.000021
1929    0.0000

The most popular cars are from 2000 and 2005, accounting for about 13% off all listings. Within the top 10, all of the cars are younger than 20 years. However, surprisingly all of them are also 9 years or older. 

In [24]:
brands = autos['brand'].value_counts(normalize=True)
print(brands[:20].sum())

0.9280224929709465


We're going to take the top 20 brands, which accounts for just over 90% of the data.   

In [25]:
brand_list = brands[:20].index
brand_dict = {}
for brand in brand_list:
    autos_brand = autos[autos["brand"] == brand]
    autos_brand_mean = round(autos_brand["price"].mean(),2)
    brand_dict[brand] = autos_brand_mean
print(sorted(brand_dict.values(), reverse=True)) 
print(brand_dict) 

[10805.08, 10460.01, 9093.65, 8485.24, 8102.54, 6334.92, 5308.54, 5231.08, 5115.33, 4757.11, 4664.89, 4296.49, 4010.77, 3699.94, 3652.1, 3542.71, 3039.47, 2876.72, 2711.8, 2395.42]
{'citroen': 3699.94, 'audi': 9093.65, 'peugeot': 3039.47, 'mercedes_benz': 8485.24, 'seat': 4296.49, 'bmw': 8102.54, 'hyundai': 5308.54, 'volkswagen': 5231.08, 'mini': 10460.01, 'opel': 2876.72, 'fiat': 2711.8, 'toyota': 5115.33, 'volvo': 4757.11, 'nissan': 4664.89, 'sonstige_autos': 10805.08, 'skoda': 6334.92, 'ford': 3652.1, 'renault': 2395.42, 'smart': 3542.71, 'mazda': 4010.77}


Of the top most represented brands in the data set, Sonstige Autos and Minis are the most expensive, while Renault and Fiat are the least expensive brands.

Now let's look at the top 6 brands (by number of cars listed. First we need to create similar dictionaries to the brand price dictionary, and then we can use the series and data frame constructors to easily compare across all three.

In [26]:
brand_list2 = brands[:6].index
price_dict = {}
mileage_dict = {}
for brand in brand_list2:
    autos_brand = autos[autos["brand"] == brand]
    # first grab mean price for this brand
    autos_brand_price = round(autos_brand["price"].mean(),2)
    price_dict[brand] = autos_brand_price
    # next grab mean mileage for this brand
    autos_brand_mileage = round(autos_brand["odometer_km"].mean(),0)
    mileage_dict[brand] = autos_brand_mileage
price_series = pd.Series(price_dict)
mileage_series = pd.Series(mileage_dict)
top_6 = pd.DataFrame(price_series, columns=['mean_price'])
top_6['mean_mileage'] = mileage_series
print(top_6)

               mean_price  mean_mileage
audi              9093.65      129288.0
bmw               8102.54      132431.0
ford              3652.10      124069.0
mercedes_benz     8485.24      130856.0
opel              2876.72      129223.0
volkswagen        5231.08      128724.0


The mileage for cars in the top 6 brands is surprisingly similar (all roughly 13K), while price varies greatly from almost 3K to 9K.

Now we're going to compelte some final cleaning. The first step is to take German words and translate to English, then map to translations for our dataset. The columns that stand out are `vehicletype`, `gearbox`, `fueltype`, and `unrepaired_damage`.

In [27]:
vehicle_type_list = autos['vehicletype'].unique()
print(vehicle_type_list)

['bus' 'limousine' 'kleinwagen' 'kombi' nan 'coupe' 'suv' 'cabrio'
 'andere']


Some of these mappings are obvious, others less so.

In [28]:
vehicle_type_dict = {'bus':'bus',
                     'limousine':'limousine',
                     'kleinwagen':'compact',
                     'kombi':'station wagon',
                     'coupe':'sports car',
                     'cabrio':'convertible',
                     'suv':'suv',
                     'andere':'other'}
autos['vehicletype'].replace(vehicle_type_dict, inplace=True)

In [29]:
gearbox_list = autos['gearbox'].unique()
print(gearbox_list)

['manuell' 'automatik' nan]


Now we repeat the process for the gearbox and other columns.

In [30]:
gearbox_dict = {'manuell':'manual',
                'automatik':'automatic'}
autos['gearbox'].replace(gearbox_dict, inplace=True)

In [31]:
fueltype_list = autos['fueltype'].unique()
print(fueltype_list)

['lpg' 'benzin' 'diesel' nan 'cng' 'hybrid' 'elektro' 'andere']


In [32]:
fueltype_dict = {'lpg':'lpg',
                 'benzin':'gas',
                 'diesel':'diesel',
                 'cng':'cng',
                 'hybrid':'hybrid',
                 'elektro':'electric',
                 'andere':'other'}
autos['fueltype'].replace(fueltype_dict, inplace=True)

Note-- chose to keep the abbreviations LPG (liquified petroleum gas, similar to propane) and CNG (compressed natural gas) as they don't have a better alternative in English.

In [33]:
unrepaired_list = autos['unrepaired_damage'].unique()
print(unrepaired_list)

['nein' nan 'ja']


In [34]:
unrepaired_dict = {'nein':'no',
                 'ja':'yes'}
autos['unrepaired_damage'].replace(unrepaired_dict, inplace=True)

Now we'll make the dates more uniform. For our purposes the times are not interesting, so we can use the just the dates and convert to an integer (e.g. `"2016-03-21"` to `20160321`.

In [35]:
autos['lastseen'] = autos['lastseen'].str[:10].str.replace('-','').astype(int)
autos['datecrawled'] = autos['datecrawled'].str[:10].str.replace('-','').astype(int)
autos['ad_created'] = autos['ad_created'].str[:10].str.replace('-','').astype(int)

Last, we need to look through the name column to see if we can have more desriptive terms that will be useful for analysis.

In [36]:
print(autos['name'].head(15))

0                      Peugeot_807_160_NAVTECH_ON_BOARD
1            BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik
2                            Volkswagen_Golf_1.6_United
3     Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...
4     Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...
5     Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...
6     VW_Golf_III_GT_Special_Electronic_Green_Metall...
7                                  Golf_IV_1.9_TDI_90PS
8                                            Seat_Arosa
9             Renault_Megane_Scenic_1.6e_RT_Klimaanlage
11                           Mercedes_A140_Motorschaden
12    Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...
13                                   Audi_A3_1.6_tuning
14    Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...
15          Corvette_C3_Coupe_T_Top_Crossfire_Injection
Name: name, dtype: object


First name seems to be name of brand, and the second name is generally the model, both which we have already. After that though is more information about the model which could be useful.

In [37]:
model_info = autos['name'].str.split('_', n=3, expand=True)
print(model_info.head(10))

            0       1        2  \
0     Peugeot     807      160   
1         BMW    740i        4   
2  Volkswagen    Golf      1.6   
3       Smart   smart   fortwo   
4        Ford   Focus        1   
5    Chrysler   Grand  Voyager   
6          VW    Golf      III   
7        Golf      IV      1.9   
8        Seat   Arosa     None   
9     Renault  Megane   Scenic   

                                                   3  
0                                   NAVTECH_ON_BOARD  
1                    4_Liter_HAMANN_UMBAU_Mega_Optik  
2                                             United  
3                   coupe_softouch/F1/Klima/Panorama  
4  6_Benzin_TÜV_neu_ist_sehr_gepflegt.mit_Klimaan...  
5         2.8_CRD_Aut.Limited_Stow´n_Go_Sitze_7Sitze  
6  GT_Special_Electronic_Green_Metallic_Reservier...  
7                                           TDI_90PS  
8                                               None  
9                                1.6e_RT_Klimaanlage  


It looks like the name is highly variable, and most of the information we've captured in other columns anyway.

One of the last things we'll do is find the most common brand/model combinations. The first thing is to create a dictionary of unique possible combinations.

In [38]:
brand_model = autos[['brand','model']]
brand_model_unique = brand_model.drop_duplicates()
print(brand_model_unique.shape[0])

330


There are 330 different combinations of brand & model in our dataset. 

In [39]:
brand_all = autos['brand'].unique()
model_all = autos['model'].unique()
brand_model = {}
for brand in brand_all:
    brand_autos = autos.loc[autos['brand']==brand, ['brand','model']]
    for model in model_all:
        brand_model_autos = brand_autos[brand_autos['model']==model]
        n_brand_model = brand_model_autos.shape[0]
        key = "{} {}".format(brand, str(model))
        if n_brand_model > 0:
            brand_model[key] = n_brand_model
sorted(brand_model, key=brand_model.get, reverse=True)[:5]

['volkswagen golf',
 'bmw 3er',
 'volkswagen polo',
 'opel corsa',
 'volkswagen passat']

The 5 most common brand/model combinations are the Volkswagen Golf, BMW 3er, Volkswagen Polo, Opel Corsa, and Opel Astra.

In [40]:
autos['odometer_km'].describe()

count     48015.000000
mean     125551.598459
std       40098.668741
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

Now we'll split the odometer values into ranges. Based on the information from .describe(), we need to split up ranges starting at 5000 to 150000. More values seem to be towards 100,000 so we'll make our first group "under 100K" and then break up the rest in 10K chunks.

In [41]:
odometer_range = {'under 100K': '',
                  '100-110K': '',
                  '110-120K': '',
                  '120-130K': '',
                  '130-140K': '',
                  '140-150K': ''}

Now that the dictionary is set up, we'll use it to find cars with mileage in a particular range and then average the cost of those cars. This way we can see the relationship between mileage and price.

In [42]:
r1 = autos.loc[(autos['odometer_km'] < 100000),'price'].mean()
odometer_range['under 100K'] = round(r1,2)
r2 = autos.loc[(autos['odometer_km'].between(99999,110000)),'price'].mean()
odometer_range['100-110K'] = round(r2,2)
r3 = autos.loc[(autos['odometer_km'].between(109999,120000)),'price'].mean()
odometer_range['110-120K'] = round(r3,2)
r4 = autos.loc[(autos['odometer_km'].between(119999,130000)),'price'].mean()
odometer_range['120-130K'] = round(r4,2)
r5 = autos.loc[(autos['odometer_km'].between(129999,140000)),'price'].mean()
odometer_range['130-140K'] = round(r5,2)
r6 = autos.loc[(autos['odometer_km'].between(139999,150000)),'price'].mean()
odometer_range['140-150K'] = round(r6,2)

print(odometer_range)

{'under 100K': 11932.6, '120-130K': 6086.21, '140-150K': 3654.0, '100-110K': 7936.03, '110-120K': nan, '130-140K': nan}


Not surprisingly, cars with the least mileage are the most expensive, and cars with the most mileage are the least expensive. Odometer value in the middle ranges follow the same pattern.

In [43]:
cars_damaged = autos.loc[autos['unrepaired_damage']=='yes','price'].mean()
cars_undamaged = autos.loc[autos['unrepaired_damage']=='no','price'].mean()
print("damaged cars are, on average: ${}".format(round(cars_damaged,2)))
print("undamaged cars are, on average: ${}".format(round(cars_undamaged,2)))

damaged cars are, on average: $2126.84
undamaged cars are, on average: $7067.21


Also not surprisingly, damaged cars are much cheaper than their undamaged counterparts.